File: neetcode/blind75/1D_DP/longest_palindromic_substring.py
# https://neetcode.io/problems/longest-palindromic-substring

class Solution:
    def longestPalindrome(self, s: str) -> str:
        
        if not s:
            return "";

        res = s[0]
        
        l, r = 0, 0
        for i in range(len(s)):
            l = i
            r = i
            # starting range, move pointers while match initial value
            while (l - 1 >= 0 and s[l - 1] == s[i]):
                l = l - 1

            while (r + 1 < len(s) and s[r + 1] == s[i]):
                r = r + 1

            # move while same char on each side
            while (l >= 0 and r < len(s) and s[l] == s[r]):
                l -= 1
                r += 1

            if ((r - (l+1)) > len(res)):
                res = s[l+1:r];

        return res

File: neetcode/blind75_JS_2/1-D-Dynamic-Programming/longest-palindromic-substring.js
// https://neetcode.io/problems/longest-palindromic-substring

/*
edge case 1: 
    if (s.length === 0) {
        return ''
    }
edge case 2:
    if (s.length === 1) {
        return s[0]
    }

create a 2D Array for the DP table of rows length = s.length and cols length = s.length. Fill with false
initial state is the same char, r === c, equates to true since a char with itself is a palindrome.

iterate rows r from n - 1 to 0
    iterate cols from r + 1 to n - 1
        if s[r] === s[c] AND (check if this char continues a palindrome, s[r + 1][c - 1] OR j - i <= 2)
        // c - r <= 2 is if the same char (diff = 0) or if match with char beside itself (diff = 1, even center), or if match with char on other side of center (diff = 2, odd center) 

- Time: O(n^2)
- Space: O(n^2)
*/

class Solution {
    /**
     * @param {string} s
     * @return {string}
     */
    longestPalindrome(s) {
        if (s.length === 0) {
            return ''
        }
        if (s.length === 1) {
            return s[0]
        }
        
        const dp = Array.from(new Array(s.length), (v) => new Array(s.length).fill(false))
        for (let r = 0; r < s.length; r ++) {
            dp[r][r] = true
        }

        const palinBounds = [0, 0]
        let palinLength = 1

        for (let r = s.length - 2; r >= 0; r --) {
            for (let c = r + 1; c < s.length; c ++) {
                if (s[r] === s[c] && (dp[r + 1][c - 1] === true || c - r <= 2)) {
                    if (c - r + 1 > palinLength) {
                        palinBounds[0] = r
                        palinBounds[1] = c
                        palinLength = c - r + 1
                    }
                    dp[r][c] = true
                }
            }
        }
        console.log(dp)
        return s.slice(palinBounds[0], palinBounds[1] + 1)
    }
}
